### PR TITLE
Implements locking Git state via Etags

### PR DESCRIPTION
**Pull request by nyarly**
The state gets an etag from the GitStateManager
(the revision of the current head).

The restul.Server will reject PUTs unless they
If-Match the current git-rev Etag.

The Etag gets set on the written state,
which the GSM checks against the head revision.

The upshot should be that it is impossible now to make an update
to the GDM unless
you are basing that update on the current revision.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: ext/storage/git_state_manager_test.go [hunk 1]
```diff
@@ -199,6 +154,33 @@ func setupManagers(t *testing.T) (clone *GitStateManager, remote *DiskStateManag
 	return gsm, dsm
 }
 
+func TestStateEtags(t *testing.T) {
+	gsm, _ := setupManagers(t)
+
+	s, err := gsm.ReadState()
+
+	if err != nil {
+		t.Errorf("Unexpected error when reading state")
```
Comment (samsalisbury) [inline ext/storage/git_state_manager_test.go]: Always helpful to include the text of captured errors for greater debuggability.

File: lib/state.go [hunk 2]
```diff
@@ -95,6 +96,32 @@ func NewState() *State {
 	}
 }
 
+// etag is a private field to ensure that we don't interfere with YAML storage,
+// hence the getter/setter
```
Comment (samsalisbury) [inline lib/state.go]: Maybe this comment should be above `etag` field?

File: server/graph.go [hunk 3]
```diff
@@ -28,5 +29,6 @@ func liveGDM(sr graph.StateReader) (*LiveGDM, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &LiveGDM{gdm.Deployments}, nil
+	etag, _ := state.GetEtag()
```
Comment (samsalisbury) [inline server/graph.go]: Should prob not swallow this error, could be confusing.

File: server/handle_gdm.go [hunk 4]
```diff
@@ -85,6 +87,10 @@ func (h *PUTGDMHandler) Exchange() (interface{}, int) {
 		return "Invalid GDM", http.StatusBadRequest
 	}
 
+	if _, got := h.Header["Etag"]; got {
+		state.SetEtag(h.Header.Get("Etag"))
```
Comment (samsalisbury) [inline server/handle_gdm.go]: Could read the header once in the initialisation?

File: util/restful/router.go [hunk 5]
```diff
@@ -170,23 +170,46 @@ func (mh *MetaHandler) writeHeaders(status int, w http.ResponseWriter, r *http.R
 	mh.statusHandler.HandleResponse(status, r, w, data)
 }
 
+var etagHeader = http.CanonicalHeaderKey("Etag")
+var contentTypeHeader = http.CanonicalHeaderKey("Content-Type")
+var contentLengthHeader = http.CanonicalHeaderKey("Content-Length")
```
Comment (samsalisbury) [inline util/restful/router.go]: These are already canonical! :P could be constants, but doesn't make much difference I guess.

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,12 +7,17 @@ and this project adheres to [Semantic Versioning](http://semver.org/)
 with respect to its command line interface and HTTP interface.
 
 ## [Unreleased](//github.com/opentable/sous/compare/0.5.16...HEAD)
+
 ### Changed
+- Etags on /gdm are generated using the current head revision of the GDM git
+  repo. Updates are checked against this revision, which should mean we have a
+  closed loop against conflicting updates.
```
Comment (samsalisbury) [inline CHANGELOG.md]: I feel like the changelog should be written from a users' perspective, e.g. "Fixed: performance issues caused by conflicting concurrent writes to the GDM taking multiple passes to resolve." or something...

File: server/graph.go (hunk 3 above)
Comment (samsalisbury) [inline server/graph.go]: Or if that's what we want, then a comment helps.

File: server/handle_gdm.go (hunk 4 above)
Comment (samsalisbury) [inline server/handle_gdm.go]: Oh, multiple fields with the same name, ignore that.

**Comment by samsalisbury** [comment]
I worry that this will generate a lot of conflicts, many of which are not relevant as they impact other manifests. It's definitely an improvement from where we were, but I think we'll need more re-tries for 'sous update' for now, until we narrow these Etags to a specific manifest...

**Pull request merged**